### ISSUE TITLE
numpy sum returns unrecognized objects unchanged

### ISSUE THREAD

**Issue by KesterTong** [numpy/numpy#8927]
When passed an object that isn't array-like, numpy.sum returns that object.  This is surprising behavior.  Maybe it is consistent with interpreting the input as a scalar, but it is surprising when the input is iterable.  E.g.
```
class A(object):
  pass

a = A()
np.sum(a)  # returns `a`
```

Note that this occurs even for iterables, e.g.

```
class A(object):
  def __getitem__(self, index):
    if index > 3:
        raise IndexError("that's enough!")

    return index

a = A()
np.sum(a)  # returns `a`
```

**Comment by eric-wieser** [comment]
As you've probably worked out, `np.sum(a)` in this case just does `np.array(a).sum()`. `np.array(a)` does not accept iterables, only sequences (meaning you need a `__len__` method too).

I think there's an issue about passing iterables to `np.array` somewhere...

**Comment by eric-wieser** [comment]
#5718 is related. In python 3, `dict_keys` has both `__item__` and `__len__`, yet does not work with `np.array`.

**Comment by KesterTong** [comment]
Thanks for the information @eric-wieser.  For this issue my main concern is not that numpy doesn't recognize the iterable, but that it silently returns the same object.

I believe that the fundamental problem is that numpy.sum does not perform any arithmetic operations when the array is 1-dimensional.  Would it be a violation of duck typing to check that the operation in question is supported (e.g. by adding the element to itself)?  That way we would get the same error as for the following code

```
np.sum([a, a])  # raises "TypeError: unsupported operand type(s) for +: 'A' and 'A'"
```

**Comment by eric-wieser** [comment]
> I believe that the fundamental problem is that numpy.sum does not perform any arithmetic operations when the array is 1-dimensional. 

I think what you mean here is _"0-length along the axis being summed"_, not _"1-dimensional"_. These also give similar results:

* np.sum([A()]
* np.sum([[A(), A()]], axis=0)

I think this is a remark that is generally applicable to `ufunc.reduce` - when working with `dtype=object`, with only one element on the reduction axis, should it first try to check if the operation is supported?

Relatedly, reducing object arrays with length 0 on the reduction axis also is weird: #8860

**Comment by eric-wieser** [comment]
I'm now thinking this is undesirable, as in the following example:
```
class AddReturnsOther(object):
    def __add__(self, other): return other
    __radd__ = __add__

np.sum([A()] + [AddReturnsOther()]*2)  # OK, returns A()
np.sum([A()] + [AddReturnsOther()]*1)  # OK, returns A()
np.sum([A()] + [AddReturnsOther()]*0)  # OK, would error under your proposal
```

**Comment by KesterTong** [comment]
Thanks for looking into this.  Given the above, I don't see a clean solution, so I think the fix is to be more aware of the rules for implicit conversion.

Perhaps the documentation on np.array could be clearer, but I'm not sure.  It currently states:

> An array, any object exposing the array interface, an object whose __array__ method returns an array, or any (nested) sequence.

I'm not sure if "sequence" is a well defined concept in Python, or if it matches what `array` actually accepts.  If not, that part could be updated.  If so, maybe link to the documentation for Python sequences (afaict a Python sequence is defined [here](https://docs.python.org/2.4/lib/typesseq.html) but this defines sequences to be the builtin sequence types)

**Comment by eric-wieser** [comment]
Having looked a bit closer, there's a special case in `ufunc.reduce` that ignores `ufunc.identity` when `dtype == object`.

Removing this case would fix your problem, but would also break code like `np.sum(np.array(["A", "B", "C"], object))`

**Comment by njsmith** [comment]
Probably deprecating the automatic coercion of unrecognized objects to object arrays would be a simpler way to prevent confusion in this case (and many, many others)

**Comment by KesterTong** [comment]
@eric-wieser would that effectively remove support for np.object in np.reduce?  I like @njsmith's idea, this seems to be the root of the problem.  In fact I didn't realize np.object type existed until I encountered this problem.

**Comment by eric-wieser** [comment]
> would that effectively remove support for np.object in np.reduce

No, it would just enforce that `np.add,reduce` only accepts objects that can be added to `0`, `np.multiply.reduce` only accepts objects that can be multiplied by `1`, etc.

I think @njsmith's solution is a much better way to address the problem you're seeing, but I think it might be handy to remove the special case  for `np.object_` in reduce.

**Comment by KesterTong** [comment]
Ah I see.  That would certainly have solved my original issue.  I don't want to influence the decision though because I have very little knowledge of how other users might be using arrays of type np.object.  Thanks for considering this so thoroughly

**Comment by charris** [comment]
> Probably deprecating the automatic coercion of unrecognized objects to object arrays would be a simpler way to prevent confusion in this case (and many, many others)

We have talked about that for years. Perhaps we should schedule that for 1.14.
